pkg/distributor/ha_tracker.go

```diff
@@ -204,9 +246,112 @@ func (c *haTracker) loop(ctx context.Context) error {
 		return true
 	})
 
+	wg.Wait()
 	return nil
 }
 
+const (
+	cleanupCyclePeriod = 30 * time.Minute
+	cleanupCycleJitter = 10 * time.Minute
+
+	// If we have received last sample for given cluster before this timeout, we will mark selected replica for deletion.
+	// If selected replica is marked for deletion for this time, it is deleted completely.
+	deletionTimeout = 30 * time.Minute
+)
+
+func (c *haTracker) cleanupOldReplicasLoop(ctx context.Context) {
+	tickerTimeout := cleanupCyclePeriod + time.Duration(rand.Int63n(cleanupCycleJitter.Nanoseconds())-cleanupCycleJitter.Nanoseconds()/2)
+
+	tick := time.NewTicker(tickerTimeout)
+	defer tick.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case t := <-tick.C:
+			c.cleanupRuns.Inc()
+			c.cleanupOldReplicas(ctx, t.Add(-deletionTimeout))
+		}
+	}
+}
+
+// Replicas marked for deletion before deadline will be deleted.
+// Replicas with last-received timestamp before deadline will be marked for deletion.
+func (c *haTracker) cleanupOldReplicas(ctx context.Context, deadline time.Time) {
+	keys, err := c.client.List(ctx, "")
+	if err != nil {
+		level.Error(c.logger).Log("msg", "cleanup: failed to list replica keys", "err", err)
```

Comment:
Maybe a warning level, considering the Cortex functionality is not affected if cleanup doesn't work?